accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +267,878 @@ also be an instance of one of the known set of subtypes. That gives us the
+    space of type `F`.
 
-    2.  Calculate `D = Lf - Rf`.
+*   **`T` is `bool`:** Two spaces of type `bool` with constant restrictions
+    `true` and `false`.
 
-    3.  If `D` is empty then `Lf` is already more a precise constraint on `L`'s
-        values than `Rf` so `R` doesn't affect this field. Add `Lf` to `fixed`.
+*   **`T` is `List<E>` for some `E` and has an arity restriction:**
 
-    4.  Else, add `D` to `changed`. (If `D` is "top", we can simply discard it.
-        There's no need to keep a field that matches every value.)
+    *The "has an arity restriction" is to ensure the space was lifted from a
+    list pattern, and not just a list constant. List constants do not help with
+    exhaustiveness because reified type arguments mean that a list constant may
+    fail to match a list that contains the same elements.*
 
-2.  Now we know which fields are affected by the subtraction and which aren't.
-    If `changed` is empty, then no fields come into play and the result is
-    simply `L`.
+    1.  Let `n` be `headMax + tailMax`. *The challenge with list patterns is
+        that we could expand them to an unbounded number of spaces: empty list,
+        list with one element, list with two elements, etc. Here, `n` represents
+        the highest individual arity we need to expand to because there is no
+        pattern that will match a list with a specific number of elements larger
+        than that.*
 
-3.  Otherwise, the result is a union of objects where each object includes one
-    of the changed fields and leaves the others alone. This is how we tile over
-    the holes created by subtracting `R`.
+    2.  For `i` from `0` to `n`, half-inclusive:
 
-    1.  For each field in `changed`, create an object with the same type as `L`
-        with all fields from `fixed`, this field from `changed` and all other
-        field names in `changed` set to their original space in `L`.
+        1.  A space with type `T` and an arity restriction with length `i` and
+            `hasRest` `false`.
 
-    2.  The result is the union of those spaces.
+    3.  And a space with type `T` and an arity restriction with length `n` and
+        `hasRest` `true`.
 
-    For example:
+    *For example, given:*
 
-    ```
-    (w: bool, x: bool, y: bool, z: bool) - (x: true, y: false, z: true)`
+    ```dart
+    switch (list) {
+      case [_, ..., _, _]:
+      case [_, _, _, ..., _]:
+    }
     ```
 
-    `fixed` is `(w: bool)`. `changed` is `(x: false, y: true, z: false)`. The
-    result is a union of:
+    *The expansion of the value list is:*
 
     ```
-    (x: false, y: bool, z: bool) // Set x to D.
-    (x: bool, y: true, z: bool) // Set y to D.
-    (x: bool, y: bool, z: false) // Set z to D.
+    List(length: 0, hasRest: false)
+    List(length: 1, hasRest: false)
+    List(length: 2, hasRest: false)
+    List(length: 3, hasRest: false)
+    List(length: 4, hasRest: false)
+    List(length: 4, hasRest: true)
```

Comment:
Shouldn't it be `List(length: 5, hasRest: true)` here? `headMax` is `3` and `tailMax` is `2`, so `n==5` here

Reply:
Oops, yes.